File: 2025/src/bin/day9/day9_p1.rs
use std::fs;
use std::cmp::max;

fn main() {
    let contents = fs::read_to_string("src/bin/day9/data.txt")
        .expect("Should have been able to read the file");

    let mut largest_area: i64 = 0;
    let mut points: Vec<Vec<i64>> = Vec::new();
    for line in contents.lines() {
        let point = line.split(',').map(str::trim).map(str::parse::<i64>).collect::<Result<Vec<i64>,_>>().unwrap();
        points.push(vec![point[0], point[1]]);
    }
    for (i, point1) in points.iter().enumerate() {
        for (j, point2) in points.iter().enumerate() {
            if i <= j {
                continue
            }
            let area = ((point1[0]-point2[0]).abs()+1)*((point1[1]-point2[1]).abs()+1);
            largest_area = max(largest_area, area);
        }
    }
    
    println!("{}", largest_area);
}

File: 2022/day9/day9_p1.py
def main():
    with open("day9/data.txt", "r") as f:
        contents = f.read()

    H = (0, 0)
    T = (0, 0)
    positions = set((0,0))
    moves = {
        'L': (-1, 0),
        'R': (1, 0),
        'U': (0, 1),
        'D': (0, -1),
    }

    for line in contents.splitlines():
        dir, n = line.split(' ')
        d = moves[dir]
        for i in range(int(n)):
            prev_H = H
            H = (H[0]+d[0], H[1]+d[1])
            if abs(H[0]-T[0]) > 1 or abs(H[1]-T[1]) > 1:
                T = prev_H
                positions.add(T)
    print(len(positions))


if __name__ == "__main__":
    main()
